### ISSUE TITLE
Standard deviation crashes Eve; doesn't allow restart

### ISSUE THREAD

**Issue by rschroll** [witheve/Eve#204]
I've had two errors when trying to use the standard deviation function.  The first occurred when I tried to use it without chunking the data:

```
thread '<main>' panicked at 'index out of bounds: the len is 1 but the index is 1', ../src/libcollections/vec.rs:1263
stack backtrace:
   1:     0x7f15f69a38ce - sys::backtrace::write::h7f613ce09e7b1d1d5As
   2:     0x7f15f69a7166 - panicking::on_panic::he6322b413efd100fGFx
   3:     0x7f15f699669e - rt::unwind::begin_unwind_inner::h16a4850d896eee23R6w
   4:     0x7f15f69970c6 - rt::unwind::begin_unwind_fmt::hc997fb3d79d46a00X5w
   5:     0x7f15f69a6b36 - rust_begin_unwind
   6:     0x7f15f69d7794 - panicking::panic_fmt::hbcb2030cdfd213bbTgC
   7:     0x7f15f69d4417 - panicking::panic_bounds_check::hb124856f262b042eZfC
   8:     0x7f15f6645334 - vec::Vec<T>.Index<usize>::index::h8037228215804762499
                        at ../src/libcollections/vec.rs:1263
   9:     0x7f15f665206e - view::join_step::closure.19696
                        at src/view.rs:178
  10:     0x7f15f6651f88 - iter::Iterator::all::h3754851288281946283
                        at ../src/libcore/iter.rs:637
  11:     0x7f15f664bcdb - view::join_step::ha252f10e16365937Msb
                        at src/view.rs:177
  12:     0x7f15f664b852 - view::join_step::ha252f10e16365937Msb
                        at src/view.rs:157
  13:     0x7f15f664b852 - view::join_step::ha252f10e16365937Msb
                        at src/view.rs:157
  14:     0x7f15f664bd49 - view::join_step::ha252f10e16365937Msb
                        at src/view.rs:182
  15:     0x7f15f664b852 - view::join_step::ha252f10e16365937Msb
                        at src/view.rs:157
  16:     0x7f15f6652c69 - view::View::run::h782c809da55f9bd3myb
                        at src/view.rs:226
  17:     0x7f15f6671da4 - flow::Flow::recalculate::h822a4f7b8118f7c2DSb
                        at src/flow.rs:165
  18:     0x7f15f6676a16 - flow::Flow::quiesce::hd82d1002a53b941e7Wb
                        at src/flow.rs:198
  19:     0x7f15f67dba5a - server::handle_event::h09e10dc5e4cfe402Ibl
                        at src/server.rs:214
  20:     0x7f15f67deeab - server::run::hf62aec871edbc828Khl
                        at src/server.rs:285
  21:     0x7f15f660242a - main::haf3dba23e5e43d0aiaa
                        at src/bin/server.rs:11
  22:     0x7f15f69a6a8d - __rust_try
  23:     0x7f15f69a8e2a - rt::lang_start::h996e83640adddea1BAx
  24:     0x7f15f66070fb - main
  25:     0x7f15f57c8ec4 - __libc_start_main
  26:     0x7f15f66022e8 - <unknown>
  27:                0x0 - <unknown>
An unknown error occurred
```
I've been having trouble reproducing this error.

The second occurred after chunking the data:
```
thread '<main>' panicked at 'Found a NaN', ../src/libcore/option.rs:331
stack backtrace:
   1:     0x7f454cefa8ce - sys::backtrace::write::h7f613ce09e7b1d1d5As
   2:     0x7f454cefe166 - panicking::on_panic::he6322b413efd100fGFx
   3:     0x7f454ceed69e - rt::unwind::begin_unwind_inner::h16a4850d896eee23R6w
   4:     0x7f454ceee0c6 - rt::unwind::begin_unwind_fmt::hc997fb3d79d46a00X5w
   5:     0x7f454cefdb36 - rust_begin_unwind
   6:     0x7f454cf2e794 - panicking::panic_fmt::hbcb2030cdfd213bbTgC
   7:     0x7f454cb66835 - option::Option<T>::expect::h9876372140545173510
                        at ../src/libcore/macros.rs:28
   8:     0x7f454cb666f7 - value::Value.Ord::cmp::h3e41548deb63d5d8iwa
                        at src/value.rs:45
   9:     0x7f454cb7c657 - cmp::impls::_&'a A.Ord::cmp::h3113803495262884271
                        at ../src/libcore/cmp.rs:592
  10:     0x7f454cb7c57e - iter::order::cmp::h12146281612936117754
                        at ../src/libcore/iter.rs:3220
  11:     0x7f454cb7c47e - slice::_[T].Ord::cmp::h9449930615456791927
                        at ../src/libcore/slice.rs:1527
  12:     0x7f454cb7c3fe - vec::Vec<T>.Ord::cmp::h10484112555876818300
                        at ../src/libcollections/vec.rs:1535
  13:     0x7f454cb89c9e - relation::Relation::change_from::h2000949c2005cd8cG1a
                        at src/relation.rs:120
  14:     0x7f454cbc917a - flow::Flow::recalculate::h822a4f7b8118f7c2DSb
                        at src/flow.rs:176
  15:     0x7f454cbcda16 - flow::Flow::quiesce::hd82d1002a53b941e7Wb
                        at src/flow.rs:198
  16:     0x7f454cd32a5a - server::handle_event::h09e10dc5e4cfe402Ibl
                        at src/server.rs:214
  17:     0x7f454cd35eab - server::run::hf62aec871edbc828Khl
                        at src/server.rs:285
  18:     0x7f454cb5942a - main::haf3dba23e5e43d0aiaa
                        at src/bin/server.rs:11
  19:     0x7f454cefda8d - __rust_try
  20:     0x7f454ceffe2a - rt::lang_start::h996e83640adddea1BAx
  21:     0x7f454cb5e0fb - main
  22:     0x7f454bd1fec4 - __libc_start_main
  23:     0x7f454cb592e8 - <unknown>
  24:                0x0 - <unknown>
An unknown error occurred
```
I can reliably reproduce this error by loading the artists.eve example file, chunking the "Top Artists" data set, and connecting the standard deviation function to the rank.  (Yes, this is a meaningless calculation....)

In both cases, Eve fails when you try to restart it, due to an error in the autosave file:
```
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: SyntaxError("trailing characters", 1, 147726)', ../src/libcore/result.rs:732
stack backtrace:
   1:     0x7fc2c19cd8ce - sys::backtrace::write::h7f613ce09e7b1d1d5As
   2:     0x7fc2c19d1166 - panicking::on_panic::he6322b413efd100fGFx
   3:     0x7fc2c19c069e - rt::unwind::begin_unwind_inner::h16a4850d896eee23R6w
   4:     0x7fc2c19c10c6 - rt::unwind::begin_unwind_fmt::hc997fb3d79d46a00X5w
   5:     0x7fc2c19d0b36 - rust_begin_unwind
   6:     0x7fc2c1a01794 - panicking::panic_fmt::hbcb2030cdfd213bbTgC
   7:     0x7fc2c17e9766 - result::Result<T, E>::unwrap::h16748058307513226779
                        at ../src/libcore/macros.rs:28
   8:     0x7fc2c17e8391 - server::load::hf7ead53dd82576eev4k
                        at src/server.rs:134
   9:     0x7fc2c180821e - server::run::hf62aec871edbc828Khl
                        at src/server.rs:257
  10:     0x7fc2c162c42a - main::haf3dba23e5e43d0aiaa
                        at src/bin/server.rs:11
  11:     0x7fc2c19d0a8d - __rust_try
  12:     0x7fc2c19d2e2a - rt::lang_start::h996e83640adddea1BAx
  13:     0x7fc2c16310fb - main
  14:     0x7fc2c07f2ec4 - __libc_start_main
  15:     0x7fc2c162c2e8 - <unknown>
  16:                0x0 - <unknown>
An unknown error occurred
```
This is presumably the same sort of problem as in #202.  I have saved both of the autosave files, in case you would like to see them.

This is seen running Eve f35374c2 on Ubuntu 14.04.

**Comment by joshuafcole** [comment]
If possible, can you gist the autosaves and link to them in each of the respective issues? That'll help us figure out the root cause of the issue.

**Comment by rschroll** [comment]
Here's the autosave after the crash from trying to connect std-dev to 
un-chunked data: https://gist.github.com/rschroll/d1b7255a9ea64bb183f0

And here's the autosave after the crash due to NaNs in the std-dev 
data: https://gist.github.com/rschroll/b1e661af550eb262acd3

**Comment by jamii** [comment]
1 is a compiler bug - I'm still working on it. 2 was from bad math and is fixed. 3 was from a missing \n in the artists.eve file and is fixed.

**Comment by rschroll** [comment]
I've just discovered that I can produce a crash much like the first by 
dropping the output of a function on one of its inputs.  Moreover, it 
produces an autosave file that causes the same error when Eve in 
resumed: https://gist.github.com/rschroll/be8a8509e31fe23a77db

I can file this as a separate bug if it's actually a separate problem.  
But given the problem I'm having reproducing the first crash, it's 
possible that this was actually the problem, triggered by mistake after 
I got the warning about the data not being chunked.
